tests/zquantum/core/wip/estimators/estimation_test.py

```diff
@@ -0,0 +1,198 @@
+            circuit.to_unitary().conj().T
+            @ qubit_operator_sparse(qubit_operator)
+            @ circuit.to_unitary()
+        )
+        assert np.allclose(target_unitary.todense(), transformed_unitary)
+
+    def test_get_context_selection_circuit_diagonal(self):
+        term = QubitOperator("Z2 Z4")
+        circuit, ising_operator = get_context_selection_circuit(term)
+        assert len(circuit.gates) == 0
+        assert ising_operator == change_operator_type(term, IsingOperator)
+
+    def test_get_context_selection_circuit_for_group(self):
+        group = QubitOperator("X0 Y1") - 0.5 * QubitOperator((1, "Y"))
+        circuit, ising_operator = get_context_selection_circuit_for_group(group)
+
+        # Need to convert to QubitOperator in order to get matrix representation
+        qubit_operator = change_operator_type(ising_operator, QubitOperator)
+
+        target_unitary = qubit_operator_sparse(group)
+        transformed_unitary = (
+            circuit.to_unitary().conj().T
+            @ qubit_operator_sparse(qubit_operator)
+            @ circuit.to_unitary()
+        )
+
+        assert np.allclose(target_unitary.todense(), transformed_unitary)
+
+    @pytest.fixture()
+    def frame_operators(self):
+        operators = [
+            2.0 * IsingOperator((1, "Z")) * IsingOperator((2, "Z")),
+            1.0 * IsingOperator((3, "Z")) * IsingOperator((0, "Z")),
+            -1.0 * IsingOperator((2, "Z")),
+        ]
+
+        return operators
+
+    @pytest.mark.parametrize(
+        "n_samples, target_n_samples_list",
+        [
+            (100, [100, 100, 100]),
+            (17, [17, 17, 17]),
+        ],
+    )
+    def test_uniform_shot_allocation(
+        self,
+        frame_operators,
+        n_samples,
+        target_n_samples_list,
+    ):
+        allocate_shots = uniform_shot_allocation(n_samples)
+        estimation_tasks = [
+            EstimationTask(operator, None, None) for operator in frame_operators
+        ]
+
+        new_estimation_tasks = allocate_shots(estimation_tasks)
+
+        for task, target_n_samples in zip(new_estimation_tasks, target_n_samples_list):
+            assert task.number_of_shots == target_n_samples
+
+    @pytest.mark.parametrize(
+        "total_n_shots, prior_expectation_values, target_n_samples_list",
+        [
+            (400, None, [200, 100, 100]),
+            (400, ExpectationValues(np.array([0, 0, 0])), [200, 100, 100]),
+            (400, ExpectationValues(np.array([1, 0.3, 0.3])), [0, 200, 200]),
+        ],
+    )
+    def test_proportional_shot_allocation(
+        self,
+        frame_operators,
+        total_n_shots,
+        prior_expectation_values,
+        target_n_samples_list,
+    ):
+        allocate_shots = proportional_shot_allocation(
+            total_n_shots, prior_expectation_values
+        )
+        estimation_tasks = [
+            EstimationTask(operator, None, None) for operator in frame_operators
```

Comment:
I believe this would cause mypy complaints - passing `None` to non-optional fields. Do I understand correctly is that the `None`s here are because the fields don't matter for the use case in this test? If such use cases exist shouldn't we change the `EstimationTask`, e.g. by making the 2nd & 3rd field optional?

Reply:
Thanks, testing using data that's close to the data that would be used in production is usually a good thing IMO - just like in ML, this guards us against unknown bias 🙂